AppActionVisualizer/app/src/main/java/com/example/appactionvisualizer/ui/activity/dashboard/DeepLinkListActivity.java

```diff
@@ -1,69 +1,413 @@
+      int score = StringUtils.matchScore(name.split(WHITESPACE), sentence.split(WHITESPACE));
+      // Build a fulfillment containing the found url
+      UrlTemplate builtUrl =
+          UrlTemplate.newBuilder()
+              .setTemplate(entity.getStructValue().getFieldsOrThrow(ENTITY_URL).getStringValue())
+              .build();
+      FulfillmentOption builtFulfillment =
+          FulfillmentOption.newBuilder().setUrlTemplate(builtUrl).build();
+      if (scoreMap.get(score) == null) scoreMap.put(score, new ArrayList<AppFulfillment>());
+      scoreMap.get(score).add(new AppFulfillment(appAction, action, builtFulfillment));
+    }
+    return getMatchedFromScore(scoreMap, action.getIntentName());
+  }
+
+  /**
+   * @param scoreMap score map of each deep link
+   * @param actionName action name
+   * @return recommended result map
+   */
+  private Map<String, List<AppFulfillment>> getMatchedFromScore(
+      TreeMap<Integer, List<AppFulfillment>> scoreMap, String actionName) {
+    // We only need the last entry which has the max score
+    Map.Entry<Integer, List<AppFulfillment>> entry = scoreMap.lastEntry();
+    // If score is only 1, we don't need this entry
+    if (entry == null || entry.getKey() <= 1) {
+      return null;
+    }
+    Map<String, List<AppFulfillment>> matched = new HashMap<>();
+    matched.put(actionName, entry.getValue());
+    return matched;
+  }
+
+  /**
+   * @param allActions The actions pool to be selected, key is the action name.
+   * @param inputWords all the words of user input.
+   * @return the most likely action(s) among all actions.
+   */
+  private Map<String, List<AppFulfillment>> getActionFromWords(
+      Map<String, List<AppFulfillment>> allActions, String[] inputWords) {
+    // Has to be tree map since we wanna a ordered score from low to high.
+    TreeMap<Integer, List<String>> scoresMap = new TreeMap<>();
+    // Iterate over all actions to find the most likely one.
+    for (Map.Entry<String, List<AppFulfillment>> entry : allActions.entrySet()) {
+      String actionName = entry.getKey();
+      String[] tobeMatched = actionNameWords.get(actionName);
+      if (tobeMatched == null || tobeMatched.length == 0) {
+        continue;
+      }
+      // Compute score to get a priority of this action.
+      int score = StringUtils.matchScore(tobeMatched, inputWords);
+      if (scoresMap.get(score) == null) {
+        scoresMap.put(score, new ArrayList<String>());
+      }
+      scoresMap.get(score).add(actionName);
+    }
+    List<String> matchedActions = scoresMap.lastEntry().getValue();
+    // Keep the map order as insertion order
+    Map<String, List<AppFulfillment>> matched = new HashMap<>();
+    for (String actionName : matchedActions) {
+      matched.put(actionName, allActions.get(actionName));
+    }
+    return matched;
+  }
+
+  /**
+   * @param appAction an app action to be parsed
+   * @return Construct an data map from single AppAction.
```

Comment:
maybe here to explain what the map key represents in the java doc

Reply:
will add comments